protocols/v2/roles-logic-sv2/src/channel_management/extended_channel_factory.rs

```diff
@@ -0,0 +1,248 @@
+//! # Extended Channel Factory
+//!
+//! A factory for creating and managing [`ExtendedChannel`]s.
+
+use crate::{
+    channel_management::{
+        extended_channel::ExtendedChannel, ShareValidationError, ShareValidationResult,
+    },
+    mining_sv2::{
+        ExtendedExtranonce as ExtendedExtranonceFactory, ExtendedExtranonceError,
+        OpenExtendedMiningChannel, SetCustomMiningJob, SetNewPrevHash, SubmitSharesExtended,
+        Target, UpdateChannel,
+    },
+    template_distribution_sv2::{NewTemplate, SetNewPrevHash as SetNewPrevHashTDP},
+    utils::{hash_rate_to_target, Id as IdFactory},
+};
+use binary_sv2::U256;
+use core::ops::Range;
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+};
+
+/// A factory for creating and managing [`ExtendedChannel`]s.
+///
+/// Allows the user to:
+/// - Create channels
+/// - Remove channels
+/// - Manage channel states upon receipt of:
+///   - [`OpenExtendedMiningChannel`] (create new channel)
+///   - [`UpdateChannel`] (update channel nominal hashrate and target)
+///   - [`NewTemplate`] (update channel template)
+///   - [`SetNewPrevHash`] or [`SetNewPrevHashTDP`] (update channel chain tip)
+///   - [`SetCustomMiningJob`] (add new custom mining job)
+///   - [`SubmitSharesExtended`] (validate shares)
+///
+/// Only suitable for mining servers, not clients.
+pub struct ExtendedChannelFactory {
+    channel_id_factory: Arc<RwLock<IdFactory>>,
+    extended_channel_factory: Arc<RwLock<HashMap<u32, Arc<RwLock<ExtendedChannel>>>>>,
+    extended_extranonce_factory: Arc<RwLock<ExtendedExtranonceFactory>>,
+    share_batch_size: usize,
+    expected_share_per_minute_per_channel: f64,
+    version_rolling_allowed: bool,
+}
+
+impl ExtendedChannelFactory {
+    pub fn new(
+        extended_extranonce_range_0: Range<usize>,
+        extended_extranonce_range_1: Range<usize>,
+        extended_extranonce_range_2: Range<usize>,
+        additional_coinbase_script_data: Option<Vec<u8>>,
+        past_jobs_capacity: usize,
+        share_batch_size: usize,
+        expected_share_per_minute_per_channel: f64,
+        version_rolling_allowed: bool,
+    ) -> Result<Self, ExtendedChannelFactoryError> {
+        let extended_extranonce = ExtendedExtranonceFactory::new(
+            extended_extranonce_range_0,
+            extended_extranonce_range_1,
+            extended_extranonce_range_2,
+            additional_coinbase_script_data.clone(),
+        )
+        .map_err(|e| ExtendedChannelFactoryError::ExtendedExtranonceError(e))?;
+
+        Ok(Self {
+            channel_id_factory: Arc::new(RwLock::new(IdFactory::new())),
+            extended_channel_factory: Arc::new(RwLock::new(HashMap::new())),
+            extended_extranonce_factory: Arc::new(RwLock::new(extended_extranonce)),
+            share_batch_size,
+            expected_share_per_minute_per_channel,
+            version_rolling_allowed,
+        })
+    }
+
+    pub async fn remove_channel(
```

Comment:
```suggestion
    pub  fn remove_channel(
```

Reply:
here and elsewhere in this struct the async redundant, we are not using any `await`